ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorAnnexXtextUtil.java

```diff
@@ -0,0 +1,191 @@
+	}
+
+	/**
+	 * Find offset of character in string that is not commented out
+	 */
+	static int findUncommentedChar(final String str, final char delim) {
+		final PeekingIterator<Character> charPeekingIt = Iterators
+				.peekingIterator(str.chars().mapToObj(e -> (char) e).collect(Collectors.toList()).iterator());
+		for (int offset = 0; charPeekingIt.hasNext(); offset++) {
+			final Character c = charPeekingIt.next();
+			if (c == delim) {
+				return offset;
+			} else if (c == '-' && charPeekingIt.peek() == '-') {
+				for (offset = offset + 1; charPeekingIt.hasNext(); offset++) {
+					final Character tmp = charPeekingIt.next();
+					if (tmp == '\n') {
+						break;
+					}
+				}
+			}
+		}
+
+		throw new RuntimeException("Cannot find terminating character");
+	}
+
+	/**
+	 * Find offset of a pair characters in string that is not commented out
+	 */
+	static int findUncommentedCharPair(final String str, final Pair<Character, Character> charsToMatch) {
+		final PeekingIterator<Character> charPeekingIt = Iterators
+				.peekingIterator(str.chars().mapToObj(e -> (char) e).collect(Collectors.toList()).iterator());
+		for (int offset = 0; charPeekingIt.hasNext(); offset++) {
+			final Character c = charPeekingIt.next();
+			if (c == charsToMatch.getKey() && charPeekingIt.peek() == charsToMatch.getValue()) {
+				return offset + 2;
+			} else if (c == '-' && charPeekingIt.peek() == '-') {
+				for (offset = offset + 1; charPeekingIt.hasNext(); offset++) {
+					final Character tmp = charPeekingIt.next();
+					if (tmp == '\n') {
+						break;
+					}
+				}
+			}
+		}
+
+		throw new RuntimeException("Cannot find character sequence " + charsToMatch.toString());
+	}
+
+	static class UndoRedoHelper extends KeyAdapter implements ExtendedModifyListener {
+		private class UndoRedoStack {
+			private final static int MAX_STACK_SIZE = 50;
+			private final Stack<ExtendedModifyEvent> undo;
+			private final Stack<ExtendedModifyEvent> redo;
+
+			public UndoRedoStack() {
+				undo = new Stack<ExtendedModifyEvent>();
+				redo = new Stack<ExtendedModifyEvent>();
+			}
+
+			public void pushUndo(final ExtendedModifyEvent undoEvent) {
+				if (undo.size() > MAX_STACK_SIZE) {
+					undo.remove(0);
+				}
+				undo.add(undoEvent);
+			}
+
+			public void pushRedo(final ExtendedModifyEvent redoEvent) {
+				if (redo.size() > MAX_STACK_SIZE) {
+					redo.remove(0);
+				}
+				redo.add(redoEvent);
+			}
+
+			public ExtendedModifyEvent popUndo() {
+				return undo.pop();
+			}
+
+			public ExtendedModifyEvent popRedo() {
+				return redo.pop();
+			}
+
+			public void clearRedo() {
+				redo.clear();
+			}
+
+			public boolean hasUndo() {
+				return !undo.isEmpty();
+			}
+
+			public boolean hasRedo() {
+				return !redo.isEmpty();
+			}
+		}
+
+		private final StyledText styledText;
+		private final UndoRedoStack undoRedoStack;
+		private boolean isUndo;
+		private boolean isRedo;
+
+		UndoRedoHelper(final StyledText styledText) {
```

Comment:
Do not use a key listener for undo and redo. These will not be the same on all systems. Mac's will use Command instead of CTRL for example(SWT.MOD1 would cover that). Z and Y will not be the same on all system. Instead, either in this class or in the dialog itself, register a handler for the undo and redo event. For example: 

```
		IHandlerService service = PlatformUI.getWorkbench().getService(IHandlerService.class);
		service.activateHandler("org.eclipse.ui.edit.undo", new AbstractHandler() {

			@Override
			public Object execute(ExecutionEvent event) throws ExecutionException {
				System.err.println("AAAA");
				// TODO Auto-generated method stub
				return null;
			}
		});
		
// On dispose:
		//service.deactivateHandler(activation);


```